tempesta_fw/sock_srv.c

```diff
@@ -1248,262 +1476,513 @@ static int
+					    tfw_cfg_sg_def->sched_arg);
 		return 0;
 	}
-
-	tfw_cfg_schref = &tfw_cfg_schref_predict;
-	return tfw_cfgop_sched(cs, ce, &tfw_cfg_sched,
-				       tfw_cfg_schref,
-				       &tfw_cfg_sched_flags);
+	return tfw_cfgop_sched(cs, ce, &tfw_cfg_sg->parsed_sg->sched,
+				       &tfw_cfg_sg->sched_arg,
+				       &tfw_cfg_sg->sched_flags);
 }
 
 static int
 tfw_cfgop_out_sched(TfwCfgSpec *cs, TfwCfgEntry *ce)
 {
-	tfw_cfg_is_set.sched = true;
-	tfw_cfg_out_schref = &tfw_cfg_out_schref_predict;
-	return tfw_cfgop_sched(cs, ce, &tfw_cfg_out_sched,
-				       tfw_cfg_out_schref,
-				       &tfw_cfg_out_sched_flags);
+	tfw_cfg_is_set.sched = 1;
+	return tfw_cfgop_sched(cs, ce, &tfw_cfg_sg_def->parsed_sg->sched,
+				       &tfw_cfg_sg_def->sched_arg,
+				       &tfw_cfg_sg_def->sched_flags);
+}
+
+static void
+tfw_cfgop_cleanup_srv_cfg(TfwCfgSrvGroup *sg_cfg)
+{
+	if (sg_cfg->orig_sg) {
+		__tfw_sg_for_each_srv(sg_cfg->parsed_sg,
+				      tfw_sock_srv_del_conns);
+		tfw_sg_release(sg_cfg->parsed_sg);
+	}
+
+	if (sg_cfg->sched_arg)
+		kfree(sg_cfg->sched_arg);
+	list_del_init(&sg_cfg->list);
+
+	kmem_cache_free(tfw_sg_cfg_cache, sg_cfg);
+}
+
+static void
+tfw_cfgop_cleanup_srv_cfgs(void)
+{
+	TfwCfgSrvGroup *sg_cfg, *tmp;
+
+	list_for_each_entry_safe(sg_cfg, tmp, &sg_cfg_list, list)
+		tfw_cfgop_cleanup_srv_cfg(sg_cfg);
+	INIT_LIST_HEAD(&sg_cfg_list);
+
+	tfw_cfg_sg = tfw_cfg_sg_def = NULL;
 }
 
 /**
  * Clean everything produced during parsing "server" and "srv_group" entries.
  */
 static void
-tfw_clean_srv_groups(TfwCfgSpec *cs)
+tfw_cfgop_cleanup_srv_groups(TfwCfgSpec *cs)
 {
-	TfwServer *srv, *tmp;
-	static const typeof(tfw_cfg_is_set) cfg_is_set_empty = { 0 };
+	TfwSrvGroup *sg, *sg_tmp;
+	TfwServer *srv, *srv_tmp;
 
-	list_for_each_entry_safe(srv, tmp, &tfw_cfg_in_slst, list) {
-		list_del(&srv->list);
-		tfw_sock_srv_del_conns(srv);
-		tfw_server_destroy(srv);
+	tfw_cfgop_cleanup_srv_cfgs();
+
+	if (tfw_runstate_is_reconfig()) {
+		tfw_sg_for_each_srv(true, tfw_sock_srv_del_conns);
+
+		return;
 	}
-	list_for_each_entry_safe(srv, tmp, &tfw_cfg_out_slst, list) {
-		list_del(&srv->list);
+
+	/* Release active configuration. */
+	tfw_sock_srv_delete_all_conns();
+	tfw_sg_release_all();
+
+	TFW_DBG2("clean orphaned servers\n");
+	/* Release structures from previous reconfigs. */
+	list_for_each_entry_safe(srv, srv_tmp, &orphan_srvs, list) {
 		tfw_sock_srv_del_conns(srv);
 		tfw_server_destroy(srv);
 	}
+	TFW_DBG2("clean orphaned groups\n");
+	list_for_each_entry_safe(sg, sg_tmp, &orphan_sgs, list) {
+		__tfw_sg_for_each_srv(sg, tfw_sock_srv_del_conns);
+		tfw_sg_release(sg);
+	}
+	INIT_LIST_HEAD(&orphan_sgs);
+}
 
-	tfw_cfg_slstsz = tfw_cfg_out_slstsz = 0;
-	tfw_cfg_is_set = cfg_is_set_empty;
 
-	tfw_sock_srv_delete_all_conns();
-	tfw_sg_release_all();
+static int
+tfw_sock_srv_cfgstart(void)
+{
+	int r;
+
+	INIT_LIST_HEAD(&sg_cfg_list);
+	if ((r = tfw_cfgop_new_sg_cfg_default()))
```

Comment:
It seems it fixes #748. If so, please mark https://github.com/tempesta-tech/tempesta/pull/851/commits/1ab95a2a85d37f9d9e84dbcf846108d0f306563c as the fix for the issue, so it will be closed on merge.

Reply:
Actually no. Default server group is created to store various server group options as global defaults. Implicit default server group is not registered as server group available after reconfig until `cfgend()`.